pkg/controller/foo/foo_controller.go

```diff
@@ -81,6 +83,67 @@ func add(mgr manager.Manager, r reconcile.Reconciler) error {
 			return err
 		}
 	}
+
+	var mapFn = handler.ToRequestsFunc(
+		func(a handler.MapObject) []reconcile.Request {
+			var owners = a.Meta.GetOwnerReferences()
+			var requests []reconcile.Request
+			for _, owner := range owners {
+				var instance = &shipsv1beta1.Foo{}
+				if err = mgr.GetClient().Get(context.TODO(), client.ObjectKey{
+					Name:      owner.Name,
+					Namespace: a.Meta.GetNamespace(),
+				}, instance); err != nil {
+					log.Error(err, "failed to get foo instance")
+					return requests
+				}
+				log.Info("got foo", "name", instance.GetName())
+
+				var opts = &client.ListOptions{
+					Namespace: instance.GetNamespace(),
+				}
+				opts = opts.MatchingLabels(instance.Labels)
+
+				var pods = &corev1.PodList{}
+				if err := mgr.GetClient().List(context.TODO(), opts, pods); err != nil {
+					log.Error(err, "failed to get pods for job")
+					return requests
+				}
+
+				for _, po := range pods.Items {
+					log.Info("watching pod", "name", po.GetName())
+					requests = append(requests, reconcile.Request{
```

Comment:
I think you've got this a bit backwards.

AFAICT, this is the relationships in your controller:

1. Ships own Jobs
2. Jobs own Pods

The controller reconciles Ships.

So, you can watch jobs directly with the `EnqueueForOwner`.
Then, for pods, watch pods, use `EnqueueRequestsFromMapFunc`, and have your map func do the following

1. use GetControllerRef to get the owning controller for the passed in pod
2. check if that controller is a job, and return nothing if it's not
3. if it is a job, fetch that job
4. use GetControllerRef to fetch the owning controller of the job
5. If that owner is a Ship, enqueue the name/namespace of that ship, otherwise return nothing